File: Bindings/Python/test/test_doxa.py
import unittest
from PIL import Image, ImageChops
import numpy as np
import os

import sys
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..', 'dist')))

import doxapy

# Get the absolute path to the RESOURCES directory (at project root)
RESOURCES_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..', '..', 'Doxa.Test', 'Resources'))

def read_image(file, algorithm=doxapy.GrayscaleAlgorithms.QT):
    full_path = os.path.join(RESOURCES_DIR, os.path.basename(file))
    image = Image.open(full_path)

    # If already in grayscale or binary, do not convert it
    if image.mode == 'L':
        return np.array(image)
    
    # Read the color image
    rgb_image = np.array(image.convert('RGB') if image.mode not in ('RGB', 'RGBA') else image)

    # Use Doxa to convert to grayscale
    return doxapy.to_grayscale(algorithm, rgb_image)

def read_weights(file):
    full_path = os.path.join(RESOURCES_DIR, os.path.basename(file))
    return doxapy.read_weights(full_path)


class DoxaPyTests(unittest.TestCase):
    def test_binarization(self):

        # Read in a color image, transforming it into a grayscale image
        image = read_image("2JohnC1V3.ppm")

        # Create a new binary image
        binary_image = np.empty(image.shape, image.dtype)

        params = {"window": 27, "k": 0.10}
        sauvola = doxapy.Binarization(doxapy.Binarization.Algorithms.SAUVOLA)
        sauvola.initialize(image)
        sauvola.to_binary(binary_image, params)

        # Update the grayscale image to binary in-place
        doxapy.update_to_binary(doxapy.Binarization.Algorithms.SAUVOLA, image, params)

        # Ensure they are equal
        self.assertTrue((binary_image == image).all())

        # Compare against our control
        expected_image_path = os.path.join(RESOURCES_DIR, os.path.basename("2JohnC1V3-Sauvola.pbm"))
        expected_image = Image.open(expected_image_path)
        #expected_image.save("expected_image.png")
        
        output_image = Image.fromarray(binary_image)
        #output_image.save('output_image.png')

        diff = ImageChops.difference(expected_image, output_image);
        #diff.show()
        self.assertIsNone(diff.getbbox())


    def test_performance(self):

        # Setup
        image = read_image("2JohnC1V3.ppm")
        groundtruth_image = read_image("2JohnC1V3-GroundTruth.pbm")
        params = {"window": 27, "k": 0.10}
        doxapy.update_to_binary(doxapy.Binarization.Algorithms.SAUVOLA, image, params)

        # Functions under test
        performanceAll = doxapy.calculate_performance(groundtruth_image, image)
        performance = doxapy.calculate_performance_ex(groundtruth_image, image, drdm=True, psnr=True)

        # Assert All
        self.assertAlmostEqual(performanceAll.get("accuracy"), 97.671, 2)
        self.assertAlmostEqual(performanceAll.get("fm"), 93.204, 2)
        self.assertAlmostEqual(performanceAll.get("recall"), 91.3811, 2)
        self.assertAlmostEqual(performanceAll.get("precision"), 95.1025, 2)
        self.assertAlmostEqual(performanceAll.get("psnr"), 16.329, 2)
        self.assertAlmostEqual(performanceAll.get("nrm"), 0.048, 2)
        self.assertAlmostEqual(performanceAll.get("mcc"), 0.918, 2)
        self.assertAlmostEqual(performanceAll.get("drdm"), 1.9519, 3)

        # Assert Partial
        self.assertEqual(performance.get("accuracy"), None)
        self.assertEqual(performance.get("fm"), None)
        self.assertEqual(performance.get("nrm"), None)
        self.assertEqual(performance.get("mcc"), None)
        self.assertEqual(performance.get("psnr"), performanceAll.get("psnr"))
        self.assertEqual(performance.get("drdm"), performanceAll.get("drdm"))


    def test_pseudo_performance(self):

        # Setup
        image = read_image("2JohnC1V3.ppm")
        groundtruth_image = read_image("2JohnC1V3-GroundTruth.pbm")
        params = {"window": 27, "k": 0.10}
        doxapy.update_to_binary(doxapy.Binarization.Algorithms.SAUVOLA, image, params)

        # Load pseudo weights
        p_weights = read_weights("2JohnC1V3-GroundTruth_PWeights.dat")
        r_weights = read_weights("2JohnC1V3-GroundTruth_RWeights.dat")

        # Calculate performance with pseudo weights
        performance = doxapy.calculate_performance(groundtruth_image, image, p_weights, r_weights)

        # Assert pseudo metrics
        self.assertAlmostEqual(performance.get("pseudo_fm"), 93.393, 2)
        self.assertAlmostEqual(performance.get("pseudo_recall"), 92.7954, 2)
        self.assertAlmostEqual(performance.get("pseudo_precision"), 93.9983, 2)


if __name__ == '__main__':
    unittest.main()

File: Bindings/Matlab/test/TestDoxa.m
classdef TestDoxa < matlab.unittest.TestCase
    % Test suite for the Doxa Matlab bindings.

    properties
        TestData
    end

    methods (TestClassSetup)
        function loadTestData(testCase)
            % Locate test resources
            baseDir = fileparts(mfilename('fullpath'));
            resourceDir = fullfile(baseDir, '..', '..', 'Doxa.Test', 'Resources');

            % Store resource dir for weight files
            testCase.TestData.resourceDir = resourceDir;

            % Create grayscale image using Qt algorithm (matches ground truth generation)
            rgb = imread(fullfile(resourceDir, '2JohnC1V3.ppm'));
            testCase.TestData.rgb = rgb;
            testCase.TestData.grayscaleImage = Doxa.Image(rgb, Doxa.Grayscale.QT);

            % Load ground truth images
            testCase.TestData.groundTruth = Doxa.Image(fullfile(resourceDir, '2JohnC1V3-GroundTruth.pbm'));
            testCase.TestData.sauvolaGroundTruth = Doxa.Image(fullfile(resourceDir, '2JohnC1V3-Sauvola.pbm'));

            % Keep raw ground truth for comparison
            testCase.TestData.sauvolaGroundTruthArray = ...
                uint8(imread(fullfile(resourceDir, '2JohnC1V3-Sauvola.pbm'))) * 255;
        end
    end

    methods (Test)
        function testBinarize(testCase)
            % Tests the binarize function for correctness.
            binary = Doxa.binarize(Doxa.Algorithms.SAUVOLA, ...
                testCase.TestData.grayscaleImage, window=27, k=0.10);

            result = binary.toArray();

            % Verify binary output (only 0 and 255)
            testCase.verifyEqual(unique(result(:)), uint8([0; 255]));

            % Verify against ground truth
            testCase.verifyEqual(result, testCase.TestData.sauvolaGroundTruthArray, ...
                'Binarized image does not match ground truth.');
        end

        function testUpdateToBinary(testCase)
            % Tests the in-place update function.
            img = Doxa.Image(testCase.TestData.grayscaleImage.toArray());
            Doxa.updateToBinary(Doxa.Algorithms.SAUVOLA, img, window=27, k=0.10);

            testCase.verifyEqual(img.toArray(), testCase.TestData.sauvolaGroundTruthArray, ...
                'In-place binarized image does not match ground truth.');
        end

        function testCalculatePerformance(testCase)
            % Tests performance metrics match Python/C++ expected values.
            metrics = Doxa.calculatePerformance( ...
                testCase.TestData.groundTruth, ...
                testCase.TestData.sauvolaGroundTruth);

            testCase.verifyEqual(metrics.accuracy, 97.671, 'RelTol', 1e-2);
            testCase.verifyEqual(metrics.fm, 93.204, 'RelTol', 1e-2);
            testCase.verifyEqual(metrics.recall, 91.381, 'RelTol', 1e-2);
            testCase.verifyEqual(metrics.precision, 95.103, 'RelTol', 1e-2);
            testCase.verifyEqual(metrics.psnr, 16.329, 'RelTol', 1e-2);
            testCase.verifyEqual(metrics.nrm, 0.048, 'RelTol', 1e-2);
            testCase.verifyEqual(metrics.mcc, 0.918, 'RelTol', 1e-2);
            testCase.verifyEqual(metrics.drdm, 1.952, 'RelTol', 1e-2);
        end

        function testPseudoPerformance(testCase)
            % Tests pseudo-metrics with weight files.
            pWeights = Doxa.readWeights(fullfile( ...
                testCase.TestData.resourceDir, '2JohnC1V3-GroundTruth_PWeights.dat'));
            rWeights = Doxa.readWeights(fullfile( ...
                testCase.TestData.resourceDir, '2JohnC1V3-GroundTruth_RWeights.dat'));

            metrics = Doxa.calculatePerformance( ...
                testCase.TestData.groundTruth, ...
                testCase.TestData.sauvolaGroundTruth, ...
                precisionWeights=pWeights, recallWeights=rWeights);

            testCase.verifyEqual(metrics.pseudoFM, 93.393, 'RelTol', 1e-2);
            testCase.verifyEqual(metrics.pseudoRecall, 92.795, 'RelTol', 1e-2);
            testCase.verifyEqual(metrics.pseudoPrecision, 93.998, 'RelTol', 1e-2);
        end

        function testGrayscale(testCase)
            % Tests grayscale conversion matches manual Qt formula.
            rgb = testCase.TestData.rgb;
            img = Doxa.Image(rgb, Doxa.Grayscale.QT);
            result = img.toArray();

            % Compute expected using Qt formula
            r = double(rgb(:,:,1));
            g = double(rgb(:,:,2));
            b = double(rgb(:,:,3));
            expected = uint8((r * 11 + g * 16 + b * 5) / 32);

            testCase.verifyEqual(result, expected, 'AbsTol', uint8(1), ...
                'Grayscale conversion does not match Qt formula.');
        end

        function testImageFromFile(testCase)
            % Tests that loading from file matches loading from array.
            resourceDir = testCase.TestData.resourceDir;

            imgFromFile = Doxa.Image(fullfile(resourceDir, '2JohnC1V3.ppm'), Doxa.Grayscale.QT);
            imgFromArray = Doxa.Image(imread(fullfile(resourceDir, '2JohnC1V3.ppm')), Doxa.Grayscale.QT);

            testCase.verifyEqual(imgFromFile.toArray(), imgFromArray.toArray(), ...
                'File-based and array-based construction produce different results.');
        end

        function testReadWeights(testCase)
            % Tests weight file loading.
            weights = Doxa.readWeights(fullfile( ...
                testCase.TestData.resourceDir, '2JohnC1V3-GroundTruth_PWeights.dat'));

            testCase.verifyFalse(isempty(weights));
            testCase.verifyTrue(isa(weights, 'double'));
            testCase.verifyGreaterThan(numel(weights), 0);
        end

        function testImageDisplay(testCase)
            % Tests that disp works without error.
            img = testCase.TestData.grayscaleImage;
            testCase.verifyWarningFree(@() disp(img));
        end

        function testAllAlgorithmsRun(testCase)
            % Ensures all algorithms can be called without error.
            algorithms = enumeration('Doxa.Algorithms');
            img = testCase.TestData.grayscaleImage;

            for i = 1:length(algorithms)
                alg = algorithms(i);
                testCase.verifyWarningFree(@() Doxa.binarize(alg, img), ...
                    ['Algorithm ' char(alg) ' failed to run.']);
            end
        end
    end
end
